[MPS] Fix `sum` and `prod` for complex types

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115514
* __->__ #115554
* #115513
* #115512

By not force-casting dtype to float

Test plan: `python -c "import torch;print(torch.linspace(-3.0, 3.0, 50, dtype=torch.cfloat, device='mps').sqrt().sin().sum())"`

Before:
```
tensor(21.1778+0.j, device='mps:0')
```
After
```
tensor(21.1778+39.1377j, device='mps:0')
```

malfet committed:
[MPS] Fix `sum` and `prod` for complex types

By not force-casting dtype to float

Test plan: `python -c "import torch;print(torch.linspace(-3.0, 3.0, 50, dtype=torch.cfloat, device='mps').sqrt().sin().sum())"`

Before:
```
tensor(21.1778+0.j, device='mps:0')
```
After
```
tensor(21.1778+39.1377j, device='mps:0')
```

[ghstack-poisoned]

lezcano commented on aten/src/ATen/native/mps/operations/ReduceOps.mm:
```diff
@@ -220,6 +220,7 @@ static void reduction_out_mps(const Tensor& input_t,
+                 inputScalarType != kComplexFloat && inputScalarType != kComplexHalf &&
                  (inputScalarType != kLong || !macOS13_3_plus)) {
         inputCastType = getMPSDataType(kFloat);
       } else if (!is_macos_13_or_newer() && inputScalarType == kHalf) {
```
I assume you need to replicate this for `kComplexHalf`  as well?

malfet replied:
No, complex MPS support is only available on MacOS-14+, so it would error out and never reach this condition

lezcano reviewed (approved):
Fair enough

malfet commented:
@pytorchbot merge -f "Lint + MPS tests are green (and we can not test it in CI yet :/ )"

malfet commented:
@pytorchbot revert -m "Broke MacOS x86 builds" -c nosignal

malfet committed:
Update on "[MPS] Fix `sum` and `prod` for complex types"


By not force-casting dtype to float

Test plan: `python -c "import torch;print(torch.linspace(-3.0, 3.0, 50, dtype=torch.cfloat, device='mps').sqrt().sin().sum())"`

Before:
```
tensor(21.1778+0.j, device='mps:0')
```
After
```
tensor(21.1778+39.1377j, device='mps:0')
```

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Fix `sum` and `prod` for complex types"


By not force-casting dtype to float

Test plan: `python -c "import torch;print(torch.linspace(-3.0, 3.0, 50, dtype=torch.cfloat, device='mps').sqrt().sin().sum())"`

Before:
```
tensor(21.1778+0.j, device='mps:0')
```
After
```
tensor(21.1778+39.1377j, device='mps:0')
```

[ghstack-poisoned]
